libpromises/evalfunction.c

```diff
@@ -2699,6 +2699,70 @@ bool EvalContextVariablePutSpecialEscaped(EvalContext *ctx, SpecialScope scope,
     return EvalContextVariablePutSpecial(ctx, scope, lval, value, type, tags);
 }
 
+/*********************************************************************/
+
+static JsonElement* ExecJSON_Pipe(const char *cmd, JsonElement *container)
+{
+    IOData io = cf_popen_full_duplex(cmd, false, false);
+
+    if (io.write_fd == -1 || io.read_fd == -1)
+    {
+        Log(LOG_LEVEL_INFO, "An error occurred while communicating with '%s'", cmd);
+        return NULL;
+    }
+
+    Log(LOG_LEVEL_DEBUG, "Opened fds %d and %d for command '%s'.",
+        io.read_fd, io.write_fd, cmd);
+
+    // write the container to a string
+    Writer *w = StringWriter();
+    JsonWrite(w, container, 0);
+    char *container_str = StringWriterClose(w);
+
+    if (PipeWrite(&io, container_str) != strlen(container_str))
+    {
+        Log(LOG_LEVEL_VERBOSE, "Couldn't send whole container data to '%s'.", cmd);
+        free(container_str);
+        return NULL;
+    }
+
+    free(container_str);
+
+    /* We can have some error message here. */
+    Rlist *returnlist = PipeReadData(&io, 5, 5);
+
+    /* If script returns non 0 status */
+    int close = cf_pclose_full_duplex(&io);
+    if (close != EXIT_SUCCESS)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "%s returned with non zero return code: %d",
+            cmd, close);
+    }
+
+    JsonElement *returnjq = JsonArrayCreate(5);
+
+    for (const Rlist *rp = returnlist; rp != NULL; rp = rp->next)
+    {
+        const char *data = RlistScalarValue(rp);
+        JsonElement *parsed = NULL;
+        if (JsonParse(&data, &parsed) == JSON_PARSE_OK)
```

Comment:
You're right, the JSON standard specifies that whitespace includes newlines and carriage returns and we should accept them. I'll add that.